cmd/commands/dex_server.go

```diff
@@ -0,0 +1,203 @@
+/*
+Copyright 2022 The Numaproj Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package commands
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/spf13/cobra"
+	"sigs.k8s.io/yaml"
+
+	sharedtls "github.com/numaproj/numaflow/pkg/shared/tls"
+	sharedutil "github.com/numaproj/numaflow/pkg/shared/util"
+)
+
+// generate TLS and config for Dex server
+func NewDexServerInitCommand() *cobra.Command {
+	var (
+		disableTls bool
+		hostname   string
+		baseHref   string
+	)
+
+	command := &cobra.Command{
+		Use:   "dex-server-init",
+		Short: "Generate dex config and TLS certificates for Dex server",
+		RunE: func(c *cobra.Command, args []string) error {
+
+			config, err := generateDexConfigYAML(hostname, baseHref, disableTls)
+			if err != nil {
+				return err
+			}
+
+			if err := os.WriteFile("/tmp/config.yaml", []byte(config), 0644); err != nil {
+				return fmt.Errorf("failed to create config.yaml: %w", err)
+			}
+
+			if !disableTls {
+				err = createTLSCerts()
+				if err != nil {
+					return fmt.Errorf("failed to create TLS certificates: %w", err)
+				}
+			}
+
+			return nil
+		},
+	}
+	command.Flags().BoolVar(&disableTls, "disable-tls", sharedutil.LookupEnvBoolOr("NUMAFLOW_DISABLE_DEX_SERVER_TLS", false), "Whether to disable authentication and authorization, defaults to false.")
+	command.Flags().StringVar(&hostname, "hostname", sharedutil.LookupEnvStringOr("NUMAFLOW_SERVER_ADDRESS", "https://localhost:8443"), "The external address of the Numaflow server.")
+	command.Flags().StringVar(&baseHref, "base-href", sharedutil.LookupEnvStringOr("NUMAFLOW_SERVER_BASE_HREF", "/"), "Base href for Numaflow server, defaults to '/'.")
+	return command
+}
+
+func generateDexConfigYAML(hostname, baseHref string, disableTls bool) ([]byte, error) {
+
+	// check if type of connector needs redirect URI
+	// <HOSTNAME>/<base_href>/login
+	redirectURL, err := url.JoinPath(hostname, baseHref, "/login")
+	if err != nil {
+		return nil, fmt.Errorf("failed to infer redirect url from config: %v", err)
+	}
+
+	var (
+		config []byte
+		dexCfg map[string]interface{}
+	)
+
+	config, err = os.ReadFile("/cfg/config.yaml")
+	if err != nil {
+		return nil, err
+	}
+
+	err = yaml.Unmarshal(config, &dexCfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal dex.config from configmap: %v", err)
+	}
+	// issuer URL is <HOSTNAME>/dex
+	issuerURL, err := url.JoinPath(hostname, "/dex")
```

Comment:
Do we need to have `baseHref` here?

Reply:
We don't. The config just needs to follow this formatting here.

https://github.com/numaproj/numaflow/blob/f3c47ed2ac500b3ad94432112b17c0e50e8e68c3/server/apis/v1/dexauthn.go#L51-L71

Reply:
So the issuer does not really make any difference, right.

Reply:
Yes, the redirect URI is most important here.